chain/sub/bcast/consistent.go

```diff
@@ -0,0 +1,180 @@
+package bcast
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+var log = logging.Logger("sub-cb")
+
+const (
+	// GcSanityCheck determines the number of epochs that in the past
+	// that will be garbage collected from the current epoch.
+	GcSanityCheck = 5
+	// GcLookback determines the number of epochs kept in the consistent
+	// broadcast cache.
+	GcLookback = 2
+)
+
+type blksInfo struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	blks   []cid.Cid
+}
+
+type bcastDict struct {
+	// thread-safe map impl for the dictionary
+	// sync.Map accepts `any` as keys and values.
+	// To make it type safe and only support the right
+	// types we use this auxiliary type.
+	m *sync.Map
+}
+
+func (bd *bcastDict) load(key multihash.Multihash) (*blksInfo, bool) {
+	v, ok := bd.m.Load(key.String())
+	if !ok {
+		return nil, ok
+	}
+	return v.(*blksInfo), ok
+}
+
+func (bd *bcastDict) store(key multihash.Multihash, d *blksInfo) {
+	bd.m.Store(key.String(), d)
+}
+
+func (bd *bcastDict) blkLen(key multihash.Multihash) int {
+	v, ok := bd.m.Load(key.String())
+	if !ok {
+		return 0
+	}
+	return len(v.(*blksInfo).blks)
+}
+
+type ConsistentBCast struct {
+	lk    sync.RWMutex
+	delay time.Duration
+	// FIXME: Make this a slice??? Less storage but needs indexing logic.
+	m map[abi.ChainEpoch]*bcastDict
+}
+
+func newBcastDict() *bcastDict {
+	return &bcastDict{new(sync.Map)}
+}
+
+// TODO: the VRFProof may already be small enough so we may not need to use a hash here.
+// we can maybe bypass the useless computation.
+func BCastKey(bh *types.BlockHeader) (multihash.Multihash, error) {
+	k := make([]byte, len(bh.Ticket.VRFProof))
+	copy(k, bh.Ticket.VRFProof)
+	binary.PutVarint(k, int64(bh.Height))
+	return multihash.Sum(k, multihash.SHA2_256, -1)
+}
+
+func NewConsistentBCast(delay time.Duration) *ConsistentBCast {
+	return &ConsistentBCast{
+		delay: delay,
+		m:     make(map[abi.ChainEpoch]*bcastDict),
+	}
+}
+
+func cidExists(cids []cid.Cid, c cid.Cid) bool {
+	for _, v := range cids {
+		if v == c {
+			return true
+		}
+	}
+	return false
+}
+
+func (bInfo *blksInfo) eqErr() error {
+	bInfo.cancel()
+	return fmt.Errorf("different blocks with the same ticket already seen")
+}
+
+func (cb *ConsistentBCast) Len() int {
+	cb.lk.RLock()
+	defer cb.lk.RUnlock()
+	return len(cb.m)
+}
+
+func (cb *ConsistentBCast) RcvBlock(ctx context.Context, blk *types.BlockMsg) {
```

Comment:
Can we add doc comments to RcvBlock / WaitForDelivery which explain in some more depth what those methods are doing?